cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -110,84 +110,127 @@ class CircuitOperation(ops.Operation):
             with repetitions or repetition_ids.
     """
 
-    _hash: Optional[int] = dataclasses.field(default=None, init=False)
-    _cached_measurement_key_objs: Optional[AbstractSet['cirq.MeasurementKey']] = dataclasses.field(
-        default=None, init=False
-    )
-    _cached_control_keys: Optional[AbstractSet['cirq.MeasurementKey']] = dataclasses.field(
-        default=None, init=False
-    )
-    _cached_mapped_single_loop: Optional['cirq.Circuit'] = dataclasses.field(
-        default=None, init=False
-    )
-
-    circuit: 'cirq.FrozenCircuit'
-    repetitions: IntParam = 1
-    qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = dataclasses.field(default_factory=dict)
-    measurement_key_map: Dict[str, str] = dataclasses.field(default_factory=dict)
-    param_resolver: study.ParamResolver = study.ParamResolver()
-    repetition_ids: Optional[List[str]] = dataclasses.field(default=None)
-    parent_path: Tuple[str, ...] = dataclasses.field(default_factory=tuple)
-    extern_keys: FrozenSet['cirq.MeasurementKey'] = dataclasses.field(default_factory=frozenset)
-    use_repetition_ids: bool = True
-    repeat_until: Optional['cirq.Condition'] = dataclasses.field(default=None)
-
-    def __post_init__(self):
-        if not isinstance(self.circuit, circuits.FrozenCircuit):
-            raise TypeError(f'Expected circuit of type FrozenCircuit, got: {type(self.circuit)!r}')
+    def __init__(
+        self,
+        circuit: 'cirq.FrozenCircuit',
+        repetitions: int = 1,
+        qubit_map: Optional[Dict['cirq.Qid', 'cirq.Qid']] = None,
+        measurement_key_map: Optional[Dict[str, str]] = None,
+        param_resolver: Optional[study.ParamResolverOrSimilarType] = None,
+        repetition_ids: Optional[Sequence[str]] = None,
+        parent_path: Tuple[str, ...] = (),
+        extern_keys: FrozenSet['cirq.MeasurementKey'] = frozenset(),
+        use_repetition_ids: bool = True,
+        repeat_until: Optional['cirq.Condition'] = None,
+    ):
+        # These fields are purely private: they should not be accessed
+        # from outside this class.
+        self._hash = None
+        self._cached_measurement_key_objs = None
+        self._cached_control_keys = None
+        self._cached_mapped_single_loop = None
+        self._extern_keys = extern_keys
```

Comment:
Can make these cached properties instead:
```python
from cirq._compat import cached_property

...
@cached_property
def _hash(self) -> int:
    return hash(...)
```
This should simplify some code below and let us get rid of a few more `object.__setattr__` calls.

Reply:
Done. (The `__setattr__` calls were no longer needed anyhow, but it's a good change anyways)

Reply:
Didn't know this existed! There are lots of other places we do manual caching for our immutable circuit and op classes and final results. (I looked into using functools.lru_cache for those, but that caused memory leaks)

Reply:
Yes, `cached_property` is very useful (it's part of `functools` in the standard library starting in python 3.8, but we still have to use a backport for now until we drop support for 3.7).